api/src/main/java/org/xwiki/apps/googleapps/internal/GoogleAppsAuthServiceImpl.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.apps.googleapps.internal;
+
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.user.api.XWikiUser;
+import com.xpn.xwiki.user.impl.xwiki.XWikiAuthServiceImpl;
+import com.xpn.xwiki.web.XWikiRequest;
+
+import java.net.URLEncoder;
+import java.security.Principal;
+import java.util.regex.Pattern;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.servlet.http.HttpSession;
+
+import org.securityfilter.realm.SimplePrincipal;
+import org.slf4j.Logger;
+import org.xwiki.apps.googleapps.CookieAuthenticationPersistence;
+import org.xwiki.apps.googleapps.GoogleAppsAuthService;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.annotation.InstantiationStrategy;
+import org.xwiki.component.descriptor.ComponentInstantiationStrategy;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.configuration.ConfigurationSource;
+import org.xwiki.container.servlet.filters.SavedRequestManager;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.text.StringUtils;
+
+/**
+ * An authenticator that can include a negotiation with the Google Cloud (e.g. Google Drive) services.
+ * This authenticator is created, configured and maintained by the GoogleAppsScriptService.
+ * @since 3.0
+ * @version $Id$
+ */
+@Component
+@InstantiationStrategy(ComponentInstantiationStrategy.PER_LOOKUP)
+public class GoogleAppsAuthServiceImpl extends XWikiAuthServiceImpl
+        implements GoogleAppsAuthService
+{
+
+    private static final String XWIKISPACE = "XWiki.";
+
+    @Inject
+    private Logger log;
+
+    @Inject
+    private ComponentManager componentManager;
+
+    private GoogleAppsManagerImpl googleAppsManager;
+
+    void setGoogleAppsManager(GoogleAppsManagerImpl m) {
```

Comment:
You can access components in non-component code using ``Utils.getComponent()``.

Reply:
Except that this GoogleAppsAuthServiceImpl can be created in the initialize phase where Utils.getComponent would not be able to give it. Or?

Reply:
Forget about ``Utils.getComponent()``, I was thinking that ``GoogleAppsAuthServiceImpl`` can't be a component, but it can, so you should inject the ``GoogleAppsManager`` component. Now, in order to avoid a cyclic dependency between ``GoogleAppsManagerImpl`` and ``GoogleAppsAuthServiceImpl`` you can inject a ``Provider<GoogleAppsManager>`` (i.e. "lazy" inject of ``GoogleAppsManager`` in ``GoogleAppsAuthServiceImpl``).

Reply:
I would prefer the AuthService to be completely initialized before the end of the initialization of the GoogleAppsManagerImpl: Right now it is so: It is created using the component manager by the Manager, then initialized with it. If it would wait on the injection of the GoogleAppsManager, it would fail its initialization. I guess using a provider is possible but I found the apporach of using setGoogleAppsManager to be safer and to allow package-internal references which a component injection would not allow (or would only allow with casts).
paul

Reply:
@polx it feels very much like a hack. You should use injection (with ``Provider`` if there is cyclic dependency) and you shouldn't rely on the actual implementation (``GoogleAppsManagerImpl``) but on the interface (``GoogleAppsManager``). I see ``GoogleAppsManager`` is a singleton so all instances of ``GoogleAppsAuthService`` will get the right ``GoogleAppsManager`` instance (the only one) injected.

Reply:
Well. @mflorea : so it means that the idea of package encapsulation should not be taken in account here?

If yes, then I need to bring any method for use by these components within the interface. This makes it bigger although I felt we needed to diminish it. It would allow other GoogleAppsManager to appear which makes us potentially fragile.

If no (we should take package encapsulation in account) then I could still cast the GoogleAppsManager that has been injected indicating a clear sign that this authService can only work with this Manager.

Thanks for your hints.

Reply:
The need to cast to the actual implementation and to call package protected methods is usually a sign of design flaw, but there are probably valid use cases. Since you have two components ``GoogleAppsManager`` and ``GoogleAppsAuthService``, they should communicate through their defined interfaces. This is the best practice. If you have valid arguments for not doing this then OK, but it should be easy to replace / overwrite one component implementation without affecting the other (e.g. in unit tests by mocking or at runtime for clients that need custom behavior).

Reply:
Ok. I decided to keep the package view and cast.
This will allow us to split even more the classes in the future as is necessary to resolve the complexity reported by CheckStyle.
paul